File: Palindromic Substrings.go
func countSubstrings(s string) int {
    dp := make(map[int]bool)
    checkMatch := func (i, j int) bool {
        key := i << 16 | j
        if val, ok := dp[key]; ok {
            return val
        } else {
            dp[key] = s[i] == s[j]
            return dp[key]
        }
    }

    n := len(s)
    count := 0
    var checkPalindrome func(i, j int);
    checkPalindrome = func(i, j int) {
        if i >= 0 && i < n && j >= 0 && j < n && checkMatch(i, j) {
            count += 1
            checkPalindrome(i-1, j+1)
        }
    }

    for i := 0; i < n; i++ {
        checkPalindrome(i, i)
        checkPalindrome(i, i+1)
    }

    return count
}

File: Palindromic Substrings.cpp
class Solution {
public:
    int countSubstrings(string check_str) {
        s = check_str;
        n = check_str.size();
        for (int i = 0; i < n; i++) {
            checkPalindrome(i, i);
            checkPalindrome(i, i+1);
        }
        return count;
    }
private:
    int count = 0;
    int n;
    string s;
    unordered_map<int, bool> dp;

    bool checkChar(int i, int j) {
        int key = i << 16 | j;
        auto found = dp.find(key);
        if (found == dp.end()) {
            bool value = (s.at(i) == s.at(j));
            dp.insert(make_pair(key, value));
            return value;
        } else {
            return found->second;
        }
    }

    void checkPalindrome(int i, int j) {
        if (i >= 0 && j < n && checkChar(i, j)) {
            count++;
            checkPalindrome(--i, ++j);
        }
    }
};
